src/wp-includes/post.php

```diff
@@ -8175,9 +8175,17 @@ function _update_term_count_on_transition_post_status( $new_status, $old_status,
 	}
 
 	// Update counts for the post's terms.
-	foreach ( (array) get_object_taxonomies( $post->post_type ) as $taxonomy ) {
-		$tt_ids = wp_get_object_terms( $post->ID, $taxonomy, array( 'fields' => 'tt_ids' ) );
-		wp_update_term_count( $tt_ids, $taxonomy );
+	foreach ( (array) get_object_taxonomies( $post->post_type, 'objects' ) as $taxonomy ) {
+		/** This filter is documented in wp-includes/taxonomy.php */
+		$counted_statuses = apply_filters( 'update_post_term_count_statuses', array( 'publish' ), $taxonomy );
+
```

Comment:
As a related issue, can term counting be avoided if both the old and the new status are counted? cc @johnbillion 

```suggestion

		// Do not recalculate if both the old and new statuses are included in term counts.
		if ( in_array( $old_status, $counted_statuses, true ) && in_array( $new_status, $counted_statuses, true ) ) {
			continue;
		}

```